integrations/ultralytics/server/server.py

```diff
@@ -0,0 +1,267 @@
+        help="The port that the model is hosted on",
+    )
+
+    return parser.parse_args()
+
+
+def _get_grid(size: int) -> torch.Tensor:
+    # adapted from yolov5.yolo.Detect._make_grid
+    coords_y, coords_x = torch.meshgrid([torch.arange(size), torch.arange(size)])
+    grid = torch.stack((coords_x, coords_y), 2)
+    return grid.view(1, 1, size, size, 2).float()
+
+
+# Yolo V3 specific variables
+_YOLO_V3_ANCHORS = [
+    torch.Tensor([[10, 13], [16, 30], [33, 23]]),
+    torch.Tensor([[30, 61], [62, 45], [59, 119]]),
+    torch.Tensor([[116, 90], [156, 198], [373, 326]]),
+]
+_YOLO_V3_ANCHOR_GRIDS = [t.clone().view(1, -1, 1, 1, 2) for t in _YOLO_V3_ANCHORS]
+_YOLO_V3_OUTPUT_SHAPES = [80, 40, 20]
+_YOLO_V3_GRIDS = [_get_grid(grid_size) for grid_size in _YOLO_V3_OUTPUT_SHAPES]
+
+
+def arrays_to_bytes(arrays: List[numpy.array]) -> str:
+    to_return = bytearray()
+    for arr in arrays:
+        arr_dtype = bytearray(str(arr.dtype), "utf-8")
+        arr_shape = bytearray(",".join([str(a) for a in arr.shape]), "utf-8")
+        sep = bytearray("|", "utf-8")
+        arr_bytes = arr.ravel().tobytes()
+        to_return += arr_dtype + sep + arr_shape + sep + arr_bytes
+    return to_return
+
+
+def bytes_to_arrays(serialized_arr: str) -> List[numpy.array]:
+    sep = "|".encode("utf-8")
+    arrays = []
+    i_start = 0
+    while i_start < len(serialized_arr) - 1:
+        i_0 = serialized_arr.find(sep, i_start)
+        i_1 = serialized_arr.find(sep, i_0 + 1)
+        arr_dtype = numpy.dtype(serialized_arr[i_start:i_0].decode("utf-8"))
+        arr_shape = tuple(
+            [int(a) for a in serialized_arr[i_0 + 1 : i_1].decode("utf-8").split(",")]
+        )
+        arr_num_bytes = numpy.prod(arr_shape) * arr_dtype.itemsize
+        arr_str = serialized_arr[i_1 + 1 : arr_num_bytes + (i_1 + 1)]
+        arr = numpy.frombuffer(arr_str, dtype=arr_dtype).reshape(arr_shape)
+        arrays.append(arr.copy())
+
+        i_start = i_1 + arr_num_bytes + 1
+    return arrays
+
+
+def _preprocess_image(
+    img: numpy.ndarray, image_size: Tuple[int] = (640, 640)
+) -> numpy.ndarray:
+    # raw numpy image from cv2.imread -> preprocessed floats w/ shape (3, (image_size))
+    img = cv2.resize(img, image_size)
+    img = img[:, :, ::-1].transpose(2, 0, 1)
+    return img.astype(numpy.float32) / 255.0
+
+
+def _preprocess_images(
+    images: List[numpy.ndarray], image_size: Tuple[int] = (640, 640)
+) -> numpy.ndarray:
+    # list of raw numpy images -> preprocessed batch of images
+    images = [_preprocess_image(img, image_size) for img in images]
+
+    batch = numpy.stack(images, image_size)  # shape: (batch_size, 3, (image_size))
+    return numpy.ascontiguousarray(batch)
+
+
+def _pre_nms_postprocess(outputs: List[numpy.ndarray]) -> torch.Tensor:
+    # postprocess and transform raw outputs into single torch tensor
+    processed_outputs = []
+    for idx, pred in enumerate(outputs):
+        pred = torch.from_numpy(pred)
```

Comment:
do we take a decent performance hit at all when loading torch tensors from numpy? Wondering if this is a ref copy or a memory copy ultimately and therefore if we should adapt the code to run in numpy rather than torch

Reply:
Ah, I guess we're limited to torch because of the import for non_max_suppression